Alps CI Tinkering

vetschn opened the pull request:
Test Pull request to check whether the Alps CI displays anything on GitHub PRs.

vetschn committed:
Simple workflow to test whether the uenv runner works.

vetschn committed:
Print more stuff in the CI

vetschn committed:
More random changes to trigger the CI

vetschn committed:
Add configuration files for quatrex environment setup and dependencies

alexnick83 commented:
It looks like this is actually working, so we should start setting up the tests. Do we want to run all tests in cupy mode or end-to-end and selected tests for larger sizes? I suppose this depends on how much time it will take, but for now CPU tests run quite fast.

vetschn committed:
Add Dockerfile and CI configuration for quatrex build and test stages

The bare metal runners seem to not be a very good option and the uenv creation with spack is too confusing.

Let's see if docker containers work.

vetschn commented:
cscs-ci run

vetschn committed:
Switch cudss install to local installer

vetschn committed:
Reorganize cuDSS installation in Dockerfile

vetschn committed:
Try to actually run some tests

vetschn committed:
Build container with MPICH

vetschn committed:
Remove mpi build steps again

vetschn committed:
Include feedback from CSCS support.

vetschn committed:
Update the distributed reference tests to not explicitly rely on `mpiexec` / `subprocess.run()`.

vetschn committed:
Remove obsolete spack configuration and environment files

vetschn commented:
> Test Pull request to check whether the Alps CI displays anything on GitHub PRs.

A lot more ended up happening to actually make this work :sweat_smile: I was initially trying to set up the CI through `uenv`/spack and whatnot, but this was pretty confusing, many files and it was hard for me to understand what actually was happening.

In a second approach i tried with containerized CI/CD, where the cscs docs are a lot better in my opinion. There were still a few things that were not super clear imo but after some trial and error it actually runs the tests now.

This PR also includes a change in the reference tests, where previously `mpiexec` was invoked using subprocess. This is now done in a way that allows the tests to also run via slurm.

Many of the tests run through daint actually pass, but quite a few have issues. This is on one hand because in these runners MPI can apparently be initialized only once per session, so tests for the CLI, for instance, that invoke `quatrex --help` result in something like
```
[cli_0]: PMIU_write error; fd=21 buf=:cmd=init pmi_version=1 pmi_subversion=1
```
This is unfortunate but unavoidable with this setup. We will probably have to skip these explicitely by telling pytest to do so.

Other tests just don't pass on the GPU, so these will have to be fixed.

---

A couple of questions we can maybe discuss:
- Do we actually run the entire test suite or only a part of it? (for instance just the reference tests in quatrex?) Which parts have to be skipped due to constraints of the setup?
- Is it fine to run the tests on the debug queue?
- We should probably add tests for cuDSS and NCCL communication (both built in the container) and perhaps other things that should only succeed in a GPU context. How to do this best?
- We can set it up so that the built containers are pushed to a separate registry, where they can be pulled from by users. Should we do this automagically in the CI or perhaps more manually only for actual releases? (Running using the container is actually pretty neat, no conda, no modules needed, just a small `toml` file and `srun --environment=quatrex`)

vetschn committed:
CLI tests do not need to be run in a Slurm environment

vetschn commented:
cscs-ci run

CLI tests using subprocess are now skipped if a slurm environment is detected. Other points will be adressed in separate PRs (see #280)

vetschn committed:
Add multi-node check for distributed load tests

vetschn committed:
Remove unused subprocess import from test_reference.py

vetschn commented:
`qttools` tests pass now on alps in distributed and single-rank mode. quatrex has only a few failing reference tests, that we'll fix later.
